[Serializer] Allow (de)normalization of empty objects in `PropertyNormalizer`

olsavmic opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix https://github.com/symfony/symfony/issues/46280#issuecomment-1127919687
| License       | MIT
| Doc PR        | --

This PR adds optional support for the normalization of objects without properties (in the case of `PropertyNormalizer`) or objects without getters (in the case of `GetSetMethodNormalizer`). Backward compatibility is preserved since this change is disabled by default for now.

Reasoning behind this change is explained in the discussion of the linked issue (https://github.com/symfony/symfony/issues/46280)

olsavmic committed:
[Serializer] Allow (de)normalization of empty objects in PropertyNormalizer and GetSetMethodNormalizer

olsavmic commented:
@mtarld The original issue is marked as `Bug`, should I target 4.x release or 6.1? I'd say it's rather a feature.

Also, I don't see any documentation for normalizers except for https://symfony.com/doc/5.2/serializer/normalizers.html#built-in-normalizers, I guess it's self-documented enough.

And lastly, I'm still wondering if the backward compatible behavior is needed in this edge case. Or at least what about adding a deprecation notice in case the original check does not pass so that we can delete it in a subsequent version? I mean I don't mind and most people configuring their normalizers can probably find this configuration option but it would prevent the unexpected first encounter of this `feature` 😄

mtarld commented:
Yes indeed, IMO it must be a feature :slightly_smiling_face: 

I guess the documentation, code, and context builders are enough.

For the BC migration path part, it is tricky. The behavior of these normalizers is changed (and the same goes for the serializer itself), so we must trigger a deprecation IMHO.

Maybe we could leverage the `func_num_args` function as it does not count the optional args that aren't explicitly given. This will hopefully work as the `allowNormalizationOfObjectsWithoutXXX` is the last property.
Something like `if (func_num_args() < 6)` then trigger a deprecation that tells the user to explicitly give the `allowNormalizationOfObjectsWithoutXXX` as it'll default to `true` in the 7.0 release.

WDYT?

olsavmic committed:
Add deprecation notice

Update changelog

olsavmic commented:
Sounds good, I've added the deprecation, updated the default symfony bundle configuration (src/Symfony/Bundle/FrameworkBundle/Resources/config/serializer.php) and updated tests so that the deprecation is not triggered :)  

Do we want to target 6.1 or 6.2 release?

mtarld commented:
As 6.1 is in feature freeze, we must target 6.2

mtarld commented on src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php:
```diff
@@ -36,6 +41,25 @@ class GetSetMethodNormalizer extends AbstractObjectNormalizer
+        $this->allowNormalizationOfObjectsWithoutAnyGetters = $allowNormalizationOfObjectsWithoutAnyGetters;
+
+        if (\func_num_args() < 7) {
+            trigger_deprecation('symfony/serializer', '6.1', '$allowNormalizationOfObjectsWithoutAnyGetters parameter of %s::__construct() should be explicitly provided since the default value will change to `true` in symfony/serializer >=7.0', __CLASS__);
```
```suggestion
            trigger_deprecation('symfony/serializer', '6.2', '$allowNormalizationOfObjectsWithoutAnyGetters parameter of %s() should be explicitly provided since the default value will change to `true` in symfony/serializer >=7.0', __METHOD__);
```

mtarld commented on src/Symfony/Component/Serializer/CHANGELOG.md:
```diff
@@ -14,6 +14,8 @@ CHANGELOG
  * Deprecate supporting denormalization for `AbstractUid` in `UidNormalizer`, use one of `AbstractUid` child class instead
  * Deprecate denormalizing to an abstract class in `UidNormalizer`
  * Add support for `can*()` methods to `ObjectNormalizer`
+ * Add `allowNormalizationOfObjectsWithoutAnyProperties` option to `PropertyNormalizer`
```
You can create a 6.2 section and put them under it :slightly_smiling_face:

mtarld commented on src/Symfony/Component/Serializer/Normalizer/PropertyNormalizer.php:
```diff
@@ -32,6 +36,25 @@
+        $this->allowNormalizationOfObjectsWithoutAnyProperties = $allowNormalizationOfObjectsWithoutAnyProperties;
+
+        if (\func_num_args() < 7) {
+            trigger_deprecation('symfony/serializer', '6.1', '$allowNormalizationOfObjectsWithoutAnyProperties parameter of %s::__construct() should be explicitly provided since the default value will change to `true` in symfony/serializer >=7.0', __CLASS__);
```
```suggestion
            trigger_deprecation('symfony/serializer', '6.2', '$allowNormalizationOfObjectsWithoutAnyProperties parameter of %s() should be explicitly provided since the default value will change to `true` in symfony/serializer >=7.0', __METHOD__);
```

mtarld reviewed (changes requested):
Almost good for me! :slightly_smiling_face:

mtarld commented on src/Symfony/Component/Serializer/Normalizer/GetSetMethodNormalizer.php:
```diff
@@ -36,6 +41,25 @@ class GetSetMethodNormalizer extends AbstractObjectNormalizer
 {
     private static $setterAccessibleCache = [];
 
+    private $allowNormalizationOfObjectsWithoutAnyGetters;
```
```suggestion
    private bool $allowNormalizationOfObjectsWithoutAnyGetters;
```

mtarld commented on src/Symfony/Component/Serializer/Normalizer/PropertyNormalizer.php:
```diff
@@ -32,6 +36,25 @@
  */
 class PropertyNormalizer extends AbstractObjectNormalizer
 {
+    private $allowNormalizationOfObjectsWithoutAnyProperties;
```
```suggestion
    private bool $allowNormalizationOfObjectsWithoutAnyProperties;
```

mtarld reviewed (approved):
LGTM!

olsavmic commented:
Failing checks seem to be unrelated

olsavmic commented:
Anything I can do to get a second review and merge it?
